docs/deep-links.md

```diff
@@ -0,0 +1,203 @@
+- Make sure that the route exists and return it from a handler under `path` key.
+- Handler function can transform query parameters if necessary.
+- For a deeper understanding of the route definitions and their properties, [check router implementation and its types](https://github.com/MetaMask/metamask-extension/blob/main/shared/lib/deep-links/routes/route.ts#L18).
+
+## Architecture
+
+For a deeper understanding of the deep links implementation and its architecture, here is the abstracted class diagram
+that shows associations and responsibilities of the particular components in the system.
+
+```mermaid
+classDiagram
+    class DeepLinkRouter {
+        -getExtensionURL: Function
+        -getState: Function
+        +install()
+        +uninstall()
+        -handleBeforeRequest()
+        -tryNavigateTo()
+        -canSkipInterstitial()
+        -formatUrlForInterstitialPage()
+        -get404ErrorURL()
+        -redirectTab()
+    }
+
+    class DeepLink {
+        -description: string
+        -extraDescription: string
+        -route: Route
+        -title: string
+        -cta: string
+        -skipDeepLinkInterstitialChecked: boolean
+        -isLoading: boolean
+        +render()
+        -updateStateFromUrl()
+        -onRemindMeStateChanged()
+    }
+
+    class Route {
+        +pathname: string
+        +getTitle: Function
+        +handler: Function
+        +constructor(options: RouteOptions)
+    }
+
+    class ParsedDeepLink {
+        +route: Route
+        +destination: Destination
+        +signature: SignatureStatus
+    }
+
+    class Destination {
+        +path: string
+        +query: URLSearchParams
+        OR
+        +redirectTo: URL
+    }
+
+    class EventEmitter {
+        +emit()
+        +on()
+    }
+
+    class RouteRegistry {
+        +routes: Map<string, Route>
+        +addRoute(route: Route)
+    }
+
+    %% Route implementations
+    class HomeRoute {
+        +pathname: "/home"
+        +getTitle()
+        +handler()
+    }
+
+    class SwapRoute {
+        +pathname: "/swap"
+        +getTitle()
+        +handler()
+    }
+
+    class NotificationsRoute {
+        +pathname: "/notifications"
+        +getTitle()
+        +handler()
+    }
+
+    %% Define relationships
+    EventEmitter <|-- DeepLinkRouter : extends
+    DeepLinkRouter --> ParsedDeepLink : parses and creates
+    DeepLinkRouter --> Route : uses routes from
+    DeepLink --> ParsedDeepLink : displays and processes
+    DeepLink --> Route : links to route
+    ParsedDeepLink --> Route : contains
+    ParsedDeepLink --> Destination : contains
+    Route <|-- HomeRoute : implements
+    Route <|-- SwapRoute : implements
+    Route <|-- NotificationsRoute : implements
+    RouteRegistry --> Route : registers and stores
+```
+
+Deep Link Router is instantiated in service worker ([background.js](https://github.com/MetaMask/metamask-extension/blob/main/app/scripts/background.js#L757)).
```

Comment:
Not if its in Firefox where there are now service workers.

Reply:
I wrote this in a little bit different way now. Let me know if calling it just `background script` is better or we should clarify something more?